File: dc_bridge/dc_bridge_core/src/uploader/status.rs
//! Status-Record row shapes the Uploader emits (ADR-0005), preserving the Humble
//! `out_files_metrics` column set. [`StatusContext`] bundles the (group, File,
//! Destination) a row is about — the one tuple every row shares — so each row
//! constructor only states what makes that row distinct.

use serde_json::{json, Value};
use std::time::{SystemTime, UNIX_EPOCH};

use super::group::{FileGroup, FileRef};
use super::store::Storage;
use super::FileMeta;

fn unix_now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

/// The (group, File, Destination) a status row is about.
pub(super) struct StatusContext<'a> {
    group: &'a FileGroup,
    file: &'a FileRef,
    storage: &'a Storage,
    remote_path: &'a str,
}

impl<'a> StatusContext<'a> {
    pub(super) fn new(
        group: &'a FileGroup,
        file: &'a FileRef,
        storage: &'a Storage,
        remote_path: &'a str,
    ) -> Self {
        Self {
            group,
            file,
            storage,
            remote_path,
        }
    }

    /// The shared status-row fields, preserving the Humble `out_files_metrics` column set.
    fn base_row(&self) -> serde_json::Map<String, Value> {
        let mut row = serde_json::Map::new();
        row.insert("kind".into(), json!("file_status"));
        row.insert("group_name".into(), json!(self.group.group_name));
        if let Some(robot_name) = &self.group.robot_name {
            row.insert("robot_name".into(), json!(robot_name));
        }
        if let Some(robot_id) = &self.group.robot_id {
            row.insert("robot_id".into(), robot_id.clone());
        }
        row.insert("local_path".into(), json!(self.file.local_path));
        row.insert(
            "remote_path".into(),
            json!(format!(
                "{}{}",
                self.storage.url_prefix,
                self.storage.object_key(self.remote_path)
            )),
        );
        row.insert("storage_type".into(), json!(self.storage.name));
        row.insert("updated_at".into(), json!(unix_now()));
        row
    }

    pub(super) fn uploaded(&self, meta: &FileMeta) -> Value {
        let mut row = self.base_row();
        row.insert("uploaded".into(), json!(true));
        row.insert("on_filesystem".into(), json!(true));
        row.insert("deleted".into(), json!(false));
        row.insert("content_type".into(), json!(meta.content_type));
        row.insert("size".into(), json!(meta.size));
        if let Some(duration) = meta.duration {
            row.insert("duration".into(), json!(duration));
        }
        Value::Object(row)
    }

    pub(super) fn missing(&self) -> Value {
        let mut row = self.base_row();
        row.insert("uploaded".into(), json!(false));
        row.insert("on_filesystem".into(), json!(false));
        row.insert("deleted".into(), json!(false));
        Value::Object(row)
    }

    pub(super) fn deleted(&self) -> Value {
        let mut row = self.base_row();
        row.insert("uploaded".into(), json!(true));
        row.insert("on_filesystem".into(), json!(false));
        row.insert("deleted".into(), json!(true));
        Value::Object(row)
    }
}

/// ADR-0005's group completion marker — the "manifest as completion checkpoint".
pub(super) fn group_complete_row(group: &FileGroup) -> Value {
    let mut row = serde_json::Map::new();
    row.insert("kind".into(), json!("group_complete"));
    row.insert("group_name".into(), json!(group.group_name));
    if let Some(robot_name) = &group.robot_name {
        row.insert("robot_name".into(), json!(robot_name));
    }
    if let Some(robot_id) = &group.robot_id {
        row.insert("robot_id".into(), robot_id.clone());
    }
    row.insert("complete".into(), json!(true));
    row.insert("file_count".into(), json!(group.files.len()));
    row.insert(
        "files".into(),
        Value::Array(
            group
                .files
                .iter()
                .map(|f| json!({ "key": f.key, "local_path": f.local_path }))
                .collect(),
        ),
    );
    row.insert("updated_at".into(), json!(unix_now()));
    Value::Object(row)
}

File: dc_bridge/src/uploader/status.cpp
// SPDX-FileCopyrightText: 2022-2026 David Bensoussan
// SPDX-License-Identifier: MPL-2.0

#include "dc_bridge/uploader/status.hpp"

#include <chrono>
#include <set>

namespace dc_bridge::uploader::status
{

namespace
{

// Fields the rows compute themselves. A custom key naming one of these means the
// Measurement and the Uploader disagree about what the name stands for.
const std::set<std::string>& computed_fields()
{
  static const std::set<std::string> fields{
    "kind",       "group_name", "robot_id", "local_path",    "remote_path",    "storage_type",
    "updated_at", "uploaded",   "deleted",  "on_filesystem", "duration",       "content_type",
    "size",       "complete",   "files",    "file_count",    "thumbnail_path",
  };
  return fields;
}

// Record keys the rows already carry, verbatim or under a Humble-era name of their own
// (`name` → group_name, `id` → robot_id). Also dropped, but not a collision: the value is
// in the row already, so re-emitting it would only put it in a second column.
const std::set<std::string>& payload_derived_fields()
{
  static const std::set<std::string> keys{ "name", "id", "robot_name" };
  return keys;
}

// Appends the emitting Measurement's custom keys (#419). Called last, so a key that names
// a field the row already carries is dropped instead of overwriting it.
void apply_custom_keys(nlohmann::json& row, const FileGroup& group)
{
  for (const auto& [key, value] : group.custom_keys)
  {
    if (!computed_fields().count(key) && !payload_derived_fields().count(key))
    {
      row[key] = value;
    }
  }
}

double unix_now()
{
  using namespace std::chrono;
  return duration<double>(system_clock::now().time_since_epoch()).count();
}

// The shared file_status fields (Humble out_files_metrics column set).
nlohmann::json base_row(const FileGroup& group, const FileRef& file, const Storage& storage,
                        const std::string& remote_path)
{
  nlohmann::json row;
  row["kind"] = "file_status";
  row["group_name"] = group.group_name;
  if (group.robot_name)
  {
    row["robot_name"] = *group.robot_name;
  }
  if (group.robot_id)
  {
    row["robot_id"] = *group.robot_id;
  }
  row["local_path"] = file.local_path;
  row["remote_path"] = storage.url_prefix + storage.object_key(remote_path);
  row["storage_type"] = storage.name;
  row["updated_at"] = unix_now();
  return row;
}

}  // namespace

nlohmann::json uploaded_row(const FileGroup& group, const FileRef& file, const Storage& storage,
                            const std::string& remote_path, const FileMeta& meta,
                            const std::optional<std::string>& thumbnail_path)
{
  nlohmann::json row = base_row(group, file, storage, remote_path);
  row["uploaded"] = true;
  row["on_filesystem"] = true;
  row["deleted"] = false;
  row["content_type"] = meta.content_type;
  row["size"] = meta.size;
  if (meta.duration)
  {
    row["duration"] = *meta.duration;
  }
  if (thumbnail_path)
  {
    // Same shape as `remote_path` above (url_prefix + key), so both columns are read the
    // same way by a consumer.
    row["thumbnail_path"] = storage.url_prefix + storage.object_key(*thumbnail_path);
  }
  apply_custom_keys(row, group);
  return row;
}

nlohmann::json missing_row(const FileGroup& group, const FileRef& file, const Storage& storage,
                           const std::string& remote_path)
{
  nlohmann::json row = base_row(group, file, storage, remote_path);
  row["uploaded"] = false;
  row["on_filesystem"] = false;
  row["deleted"] = false;
  apply_custom_keys(row, group);
  return row;
}

nlohmann::json deleted_row(const FileGroup& group, const FileRef& file, const Storage& storage,
                           const std::string& remote_path)
{
  nlohmann::json row = base_row(group, file, storage, remote_path);
  row["uploaded"] = true;
  row["on_filesystem"] = false;
  row["deleted"] = true;
  apply_custom_keys(row, group);
  return row;
}

nlohmann::json shed_row(const FileGroup& group, const FileRef& file, const Storage& storage,
                        const std::string& remote_path)
{
  nlohmann::json row = base_row(group, file, storage, remote_path);
  row["uploaded"] = false;
  row["on_filesystem"] = false;
  row["deleted"] = true;
  apply_custom_keys(row, group);
  return row;
}

nlohmann::json group_complete_row(const FileGroup& group)
{
  nlohmann::json row;
  row["kind"] = "group_complete";
  row["group_name"] = group.group_name;
  if (group.robot_name)
  {
    row["robot_name"] = *group.robot_name;
  }
  if (group.robot_id)
  {
    row["robot_id"] = *group.robot_id;
  }
  row["complete"] = true;
  row["file_count"] = group.files.size();
  nlohmann::json files = nlohmann::json::array();
  for (const auto& f : group.files)
  {
    files.push_back({ { "key", f.key }, { "local_path", f.local_path } });
  }
  row["files"] = std::move(files);
  row["updated_at"] = unix_now();
  apply_custom_keys(row, group);
  return row;
}

bool is_reserved_field(const std::string& name)
{
  return computed_fields().count(name) > 0;
}

}  // namespace dc_bridge::uploader::status
